### PR TITLE
Adds wrapper functions and tests

### LINKED ISSUE #25: Create wrapper function for grid optimization
This issue can be closed when a grid optimization function has been created in ``tools.py``. 
The function should 
Parameters:
- All parameters required for training an ESN
- The optimization parameters
- Verbose (print or not print)
- Timer

Returns: 
Optimal set of parameters

- [ ] Unit function to identify optimal set
- [ ] Function added
- [ ] Tests added

### PR DESCRIPTION
**Pull request by samgdotson**
This PR Closes #25 by adding wrapper functions for ESN predictions and hyperparameter optimization. Appropriate tests have been added to ``tests/`` and all pass. CircleCI should attempt to build this code and confirm this. 
I have also added a simple example notebook for training and optimizing an echo state network.

### CODE REVIEW

**Comment by samgdotson** [comment]
Closing and reopening to trigger circle ci...

**Comment by samgdotson** [comment]
CircleCI is failing because I haven't told it how to build pyESN

**Comment by samgdotson** [comment]
closing and reopening

**Comment by samgdotson** [comment]
The process was:
1. CircleCI failed because it didn't know how to get pyESN.
2. I edited the ``config.yml`` file and merged to ``arfc/main``. 
3. This didn't fix the problem immediately because CircleCI was running the configuration file *_in_* the pull request.
4. The config file in this PR need to be updated, which was accomplished by ``git pull arfc main`` and ``git push origin tools``. 
5. This triggered a new CircleCI build that successfully ran pyESN and all of the tests.

**Review by LukeSeifert — COMMENTED** [review]
Apologies for not commenting on code, I couldn't seem to get it to work. It said I needed to be on a branch to comment on it? It would let me comment on the last file, but by that point I had already written all this so I figured I would keep it all in one place.
Also, I am not too familiar with this work, so I may make unnecessary suggestions, just a warning!

**cosine_example.ipynb:**

- Consider changing t to domain on In [2] since the plot has it as domain and not t.
- In [2] plot doesn't have a title but In [15] does.
- In [6] jumps to In [15]? Not too familiar with this, unsure if intended.
- In [15] no ylabel on plot.
- In [15] goes to In [8] and then to In [16]? Again, unsure if intended.
- In [16] no ylabel.
- In [17] no ylabel.
- Should the stuff under In [11], In [12], and In [13] have Out[11], Out[12], and  Out[13] in front?
- In[11] plot has different fontsize compared to previous plots
- In[12] plot title smaller than other labels
- Can you remove the blank In [ ] box at bottom?
- Comments may be good to add.

**esn_windpower.ipynb:**

- LGTM

**optimizers.py:**

- Should xset default to None in grid_optimizer? kwargs** correspond to xset and yset.

**tests/test_optimizers.py:**

- Can there be cases of 0 variables specified or 0 sets given?
- Can there be cases of more than 2 variables/sets given?

**tests/test_tools.py:**

- LGTM

**tools.py:**

- lines 99-100, n_inputs == n_outputs == n_vars?

**Comment by samgdotson** [comment]
> Here is a general outline of things I found that may need to be changed, but everything seemed good overall to me.
> 
> **cosine_example.ipynb:**
> 
> * Consider changing t to domain on In [2] since the plot has it as domain and not t.
> * In [2] plot doesn't have a title but In [15] does.
> * In [6] jumps to In [15]? Not too familiar with this, unsure if intended.
> * In [15] no ylabel on plot.
> * In [15] goes to In [8] and then to In [16]? Again, unsure if intended.
> * In [16] no ylabel.
> * In [17] no ylabel.
> * Should the stuff under In [11], In [12], and In [13] have Out[11], Out[12], and  Out[13] in front?
> * In[11] plot has different fontsize compared to previous plots
> * In[12] plot title smaller than other labels
> * Can you remove the blank In [ ] box at bottom?
> * Comments may be good to add.

This notebook is just an example of how to use the wrapper functions I've made. The "jump" you noticed 
just means I ran cells out of order.  I've cleaned it up a little bit. In the future I might have an undergrad
clean it up even more. Possibly adding more examples as well. 


> **optimizers.py:**
> 
> * Should xset default to None in grid_optimizer? kwargs** correspond to xset and yset.

That's a great question! No, it shouldn't because users shouldn't be allowed to optimize over nothing. 
E.g. Model accuracy due to training length is independent of other hyperparameters. But model size and
its "connectedness" are strongly related. In the first case, you can optimize over just training length. In the
second case you should find the optimal set of ``n_reservoir`` and ``sparsity``. 


> **tests/test_optimizers.py:**
> 
> * Can there be cases of 0 variables specified or 0 sets given?
> * Can there be cases of more than 2 variables/sets given?

I've added a test to cover your first point. There should always be a set given. Good catch!
The second point is covered by the first test in ``test_optimizers.py``: "Too many variables given."

In the future I might add a functionality for up to three variables which would create an surface to 
minimize. Currently it's not necessary.

> **tools.py:**
> 
> * lines 99-100, n_inputs == n_outputs == n_vars?

I'm not sure what you're asking. Yes, the number of inputs should match the number of outputs. This number is defined by the shape of the input data.


Good job in this first PR! I've made some of the changes you recommend, so I'm going to send you another review request. If you think I've made adequate changes to the code, you can go ahead and "Approve" and "Merge."

**Comment by LukeSeifert** [comment]
The changes look good, I will go ahead and approve and merge!
